tv_apps/smart-home/js/edit.js

```diff
@@ -65,33 +63,34 @@
       this.isFolderReady = false;
       // The hovering card
       this._hoveringCard = null;
-      this.mainSection.dataset.mode = '';
+      this.mode = '';
     },
     /**
      * State of home app looks like this:
      * (normal state)<-->Edit mode<-->Arrange mode
      * (Only adjacant states are switchable)
      */
     toggleEditMode: function() {
-      if (this.mainSection.dataset.mode === 'edit') {
-        this.mainSection.dataset.mode = '';
+      if (this.mode === 'edit') {
+        this.mode = '';
         this.spatialNavigator.multiAdd(this.regularNavElements);
         this.spatialNavigator.multiRemove(this.editNavElements);
         this.cardScrollable.setScale();
         this.folderScrollable.setScale();
         this.cardScrollable.listElem.classList.add('exiting-edit-mode');
         this.cardManager.writeCardlistInCardStore({cleanEmptyFolder: true})
           .then(function() {
-            // Since writeCardlistInCardStore triggers card-removed event that
-            // causes re-focus on other elements in card list, we need to wait
-            // until those actions to be done before focusing editButton.
-            this.spatialNavigator.focus(this.editButton);
-            this.currentNode.classList.remove('focused');
+            // During the process of writeCardlistInCardStore,
+            // editing works is still under going
+            // and card focus would still change.
+            // So we fire 'exit-edit-mode' event
+            // after the process of writeCardlistInCardStore to let
+            // outside able to do following-up works based on final outcomes.
+            this.fire('exit-edit-mode');
           }.bind(this));
         this._concealPanel(this.currentScrollable, this.currentNode);
-        this.fire('exit-edit-mode');
       } else {
-        this.mainSection.dataset.mode = 'edit';
+        this.mode = 'edit';
```

Comment:
I guess you need to try to fix the following code: find current focusing scrollable and use it as parameter to `setCurrentScrollable`.

And at the end call `realignToReferenceElement` again if we are focusing on `folderScrollable`.